### PR TITLE
[R-package] Attempt to add group to CV using RLE [WIP]

### PR DESCRIPTION
**Pull request by Laurae2**
Attempt to fix issue #390.

Test on the following:

```r
# lgb.unloader(wipe = TRUE) # Instant environment wiping
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
train$label[train$data[, 126] == 1] <- 3
train$label <- train$label + 1
digits <- c(1000, 1999, 2001, 1513)
dtrain <- lgb.Dataset(train$data, label = train$label, group = digits, free_raw_data = FALSE)
params <- list(objective = "lambdarank",
               metric = "ncdg",
               ncdg_at = "1,2,3",
               metric_freq = 1,
               max_position = 3,
               label_gain = "0,1,3,7")
model <- lgb.cv(params,
                dtrain,
                10,
                nfold = 4,
                min_data = 1,
                learning_rate = 1,
                early_stopping_rounds = 10,
                folds = list(1:1000,
                             1001:2999,
                             3000:5000,
                             5001:6513))
```

### CODE REVIEW

**Comment by Laurae2** [comment]
@guolinke do you know why is this happening? Group is not stored the same way if it is unique, check carefully at the end of the log below:

```r
Browse[2]> setinfo(dtest, "group", rle(inverted.rle[folds[[1]]])$lengths)
<lgb.Dataset>
  Public:
    construct: function () 
    create_valid: function (data, info = list(), ...) 
    dim: function () 
    finalize: function () 
    get_colnames: function () 
    getinfo: function (name) 
    initialize: function (data, params = list(), reference = NULL, colnames = NULL, 
    save_binary: function (fname) 
    set_categorical_feature: function (categorical_feature) 
    set_colnames: function (colnames) 
    set_reference: function (reference) 
    setinfo: function (name, info) 
    slice: function (idxset, ...) 
    update_params: function (params) 
  Private:
    categorical_feature: NULL
    colnames: cap-shape=bell cap-shape=conical cap-shape=convex cap-sh ...
    free_raw_data: FALSE
    get_handle: function () 
    handle: NULL
    info: 1000
    params: list
    predictor: NULL
    raw_data: NULL
    reference: lgb.Dataset, R6
    set_predictor: function (predictor) 
    used_indices: 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 ...

Browse[2]> setinfo(dtrain, "group", rle(inverted.rle[-folds[[1]]])$lengths)
<lgb.Dataset>
  Public:
    construct: function () 
    create_valid: function (data, info = list(), ...) 
    dim: function () 
    finalize: function () 
    get_colnames: function () 
    getinfo: function (name) 
    initialize: function (data, params = list(), reference = NULL, colnames = NULL, 
    save_binary: function (fname) 
    set_categorical_feature: function (categorical_feature) 
    set_colnames: function (colnames) 
    set_reference: function (reference) 
    setinfo: function (name, info) 
    slice: function (idxset, ...) 
    update_params: function (params) 
  Private:
    categorical_feature: NULL
    colnames: cap-shape=bell cap-shape=conical cap-shape=convex cap-sh ...
    free_raw_data: FALSE
    get_handle: function () 
    handle: NULL
    info: list
    params: list
    predictor: NULL
    raw_data: NULL
    reference: lgb.Dataset, R6
    set_predictor: function (predictor) 
    used_indices: 1001 1002 1003 1004 1005 1006 1007 1008 1009 1010 1011 1 ...

Browse[2]> booster <- Booster$new(params, dtrain)
[LightGBM] [Info] Total Bins 232
[LightGBM] [Info] Number of data: 5513, number of used features: 116

Browse[2]> booster$add_valid(dtest, "valid")
Error in private$info[[name]] : subscript out of bounds

Browse[2]> dtest[['.__enclos_env__']]$private$info
group 
 1000 

Browse[2]> dtrain[['.__enclos_env__']]$private$info
$group
[1] 1999 2001 1513
```

But this "works", but is not training anything..:

```r
> model <- lgb.cv(params,
+                 dtrain,
+                 10,
+                 nfold = 2,
+                 min_data = 1,
+                 learning_rate = 1,
+                 early_stopping_rounds = 10,
+                 folds = list(1:2999,
+                              3000:6513))
[LightGBM] [Info] Total Bins 232
[LightGBM] [Info] Number of data: 3514, number of used features: 116
[LightGBM] [Info] Total Bins 232
[LightGBM] [Info] Number of data: 2999, number of used features: 116
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
[LightGBM] [Info] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] Trained a tree with leaves=1 and max_depth=1
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements.
```

Also, do you know how to add a metric? It does not seem to want to add a metric for lambdarank (I might be doing it wrong, I also tried the same as OP #390 but it did not want to provide metric during training:

```r
params <- list(objective = 'lambdarank',
               boosting_type = 'gbdt',
               label_gain = '0,1,2,3,4,5,6,7,8,9,10',
               metric = "ndcg",
               metric_freq = 1,
               is_training_metric = T,
               ndcg_at = '10',
               learning_rate = 0.1,
               verbosity = 10)
```

**Comment by guolinke** [comment]
@Laurae2 I think it is kind of R feature, not sure about it. 
if number of query is too small, it still cannot be trained. The reason is the many data will have zero grad/hess in such condition.

**Comment by Laurae2** [comment]
Closing this PR until I find a better solution.

Not sure I can find a solution to this as this is the same issue for xgboost. User has to make CV manually in xgboost.
